clang/lib/Analysis/UnsafeBufferUsage.cpp

```diff
@@ -420,25 +420,63 @@ AST_MATCHER(ArraySubscriptExpr, isSafeArraySubscript) {
   //    already duplicated
   //  - call both from Sema and from here
 
-  const auto *BaseDRE =
-      dyn_cast<DeclRefExpr>(Node.getBase()->IgnoreParenImpCasts());
-  if (!BaseDRE)
+  if (const auto *BaseDRE =
+          dyn_cast<DeclRefExpr>(Node.getBase()->IgnoreParenImpCasts())) {
+    if (!BaseDRE->getDecl())
+      return false;
+    if (const auto *CATy = Finder->getASTContext().getAsConstantArrayType(
+            BaseDRE->getDecl()->getType())) {
+      if (const auto *IdxLit = dyn_cast<IntegerLiteral>(Node.getIdx())) {
```

Comment:
It feels like this block is repeated, more or less, several times. Is there a way to factor it?

Reply:
True but the block is also trivial and I am not convinced factoring it out as a separate function is worth it or leads to better code.

Reply:
I was thinking more a lambda. I don't know LLVM's coding conventions very well, though.

Reply:
I don't think it's worth it.

Reply:
I suspect that this could have been a:
```c++
const Expr *BaseE = Node.getBase()->IgnoreParenImpCasts();
if (isa<DeclRefExpr, StringLiteral>(BaseE)) {
  if (const auto *CATy = Finder->getASTContext().getAsConstantArrayType(BaseE->getType())) {
    ...
  }
}
```
Which would also eliminate duplication. (Unless `DeclRefExpr->getDecl()->getType()` is somehow significantly different from `DeclRefExpr->getType()`.)

Another thing we can try is to simply eliminate the `isa` entirely. Like, we know that it's "an" expression, and its type is a constant-size array type. Do we really need more? Why not simply trust the type system? (It's not like the C type system has ever lied to us right? 😅)